0011-B-Distributed-Lock-building-block.md

```diff
@@ -0,0 +1,359 @@
+ 	// This may not be supported by all lock stores.
+ 	// This is a potentially destructive operation that should be used with caution.
+ 	Force bool `json:"force"`
+}
+```
+
+The main differences include:
+
+- New blocking `Lock` and `RenewLock` methods.
+- Removed the concept of "lock owners", supplied by the caller, in favor of lock IDs that are generated by the component.
+- Standard errors are returned to indicate situations such as resource already locked or resources that fails to be unlocked (or whose lock fails to be renewed) due to an invalid lock ID.
+- TTLs are pased as `time.Duration`, and components have a default TTL.
+- Added a `force` option in `Unlock`.
+
+### Runtime APIs
+
+The new APIs aim at offering a simplified user experience where developers interact with at most 3 Dapr APIs, and Dapr takes care of all background activities including renewing locks. This allows Dapr to provide additional, differentiated value to users too.
+
+We propose two possible ways to implement the new APIs, which are both proposed below, each with pros and cons. The difference lies in how the user removes the lock.
+
+**In both cases**:
+
+- Dapr renews locks automatically in background. Apps do not need to invoke a method to renew a lock explicitly. The TTL sent in requests is only used to control how frequently locks are to be renewed (lower TTLs cause locks to be renewed more frequently, but allow for quicker detection of failed hosts).
+- If the app becomes unhealthy (when app healthchecks are enabled) or if the sidecar crashes, locks are not renewed and are left to expire when the TTL ends.
+
+> About force unlocking: this is implemented primarily to recover from situations where the lock was owned by an app that is known to have crashed. This is a potentially dangerous activity, as the previous owner of the lock is not informed (at least, not immediately) that the lock has been taken away, at least not until it tries to renew or relinquish the lock.  
+> Some components may not implement forced unlocking.
+
+#### Option 1: explicit `Unlock`
+
+In this case, there are three methods:
+
+- `Lock` is used to acquire a lock (unless `immediate` is true, this blocks if the resource is already locked, until the request is canceled). The response includes a lock ID which is the identifier of the lock, and the app can use it to unlock the resource.
+- `Unlock` is invoked with the lock ID to unlock a resource.
+- `ForceUnlock` is invoked to force the removal of a lock.
+
+For gRPC:
+
+```proto
+service Dapr {
+  rpc Lock (LockRequest) returns (LockResponse);
+  rpc Unlock (UnlockRequest) returns (UnlockResponse);
+  rpc ForceUnlock (ForceUnlockRequest) returns (ForceUnlockResponse);
+}
+
+message LockRequest {
+  // When set to true, if the resource is already locked, the method returns right away with an error, and does not wait for the resource to become available.
+  bool immediate = 1;
```

Comment:
It wasn't immediately (pun not intended) clear to me what this flag did without reading the description. Could we use some variation of `wait` which might be more natural? If we want to keep the default of no wait, then perhaps `noWait`?

Reply:
Sure